### PR TITLE
WIP: security context constraints

### PR DESCRIPTION
**Pull request by pweil-**
Going to go ahead and create this WIP for SCCs so it is visible.  I am in the testing phase for the generation and strats at this point while @pmorie is working on the admission pieces.

Related PRS

1.  https://github.com/openshift/origin/pull/2544 UID allocation
1.  https://github.com/pweil-/origin/pull/18 - admission (PR to this branch)

@pmorie @smarterclayton

### CODE REVIEW

**Comment by pmorie** [comment]
@smarterclayton @pweil- Let's agree on where we are going to store the allocate UID and mcs level; once we have that I can code the rest of the admission controller.

Also, @pweil- and I were discussing yesterday that we need to annotate the pod with the SCC that validated it to perform the admission control / validation in the kubelet.  Any thoughts about that, @smarterclayton?

**Comment by liggitt** [comment]
What admission is happening on the kubelet? Shouldn't it assume the context in the pod has been validated? That's how other admission bits (quota, secret usage, etc) work

**Comment by pmorie** [comment]
That is a good point, @liggit.  @pweil- do we really need admission in the kubelet?

**Comment by pmorie** [comment]
@liggitt the admission control was going to validate that the scc that
matched to a pod still allowed everything the pod wanted to do

**Comment by derekwaynecarr** [comment]
There is no Kubelet admission control.  I am confused on this thread.  Are we just saying that we were going to have code in Kubelet that ensures the security context has the permissions it thought it had and if not fail the pod?

**Comment by pmorie** [comment]
@derekwaynecarr yep, were using 'admission control' as shorthand.

**Comment by smarterclayton** [comment]
We need restriction of pods based on node policy but it's not our primary 3.0 goal.

**Comment by smarterclayton** [comment]
> On May 31, 2015, at 1:04 PM, Paul Morie <notifications@github.com> wrote:
> 
> @smarterclayton @pweil- Let's agree on where we are going to store the allocate UID and mcs level; once we have that I can code the rest of the admission controller.
> 
My current belief is I'm going to create the following annotations:

    openshift.io/scc.sa.uid-range/<service account name>=10000-19999
    openshift.io/scc.sa.mcs/<service account name>=c50,c10

Admin can change either of these two if they want, and we'll leave it.  They can change the mcs label to be more specific and we should honor it.  

If the annotations aren't set but the config says they should be used (global config security.defaultSccMode=namespace-allocated or similar) then pod creation fails.  We may also want an annotation on the namespace that changes the assigned scc for an sa, which would then be easier - effectively an scc binding for the sa on the namespace.  On create/update we look at the binding, parameterize if it if necessary, then apply the rules.

**Comment by smarterclayton** [comment]
Going to alter my statement above slightly - going to set this:

```
openshift.io/scc.sa.uid-range=10000-19999
openshift.io/scc.sa.mcs=c50,c10
```

All service accounts in a namespace are confined to these settings.  We can do special cases (additional ranges or more labels) later.  No differentiation among service accounts.

**Comment by pweil-** [comment]
> What admission is happening on the kubelet? Shouldn't it assume the context in the pod has been validated? That's how other admission bits (quota, secret usage, etc) work

The only piece that requires kubelet information to work would be the use case of "run as non-root" but not having preallocated uids.  In this case, the user may submit a pod with no `RunAsUser` specified and be expecting the image to container a `USER` directive that the kubelet would need to check and ensure is non-root.  If that use case is not required then we're fine with not changing the kubelet's `CanRunPod` method for now.

**Comment by smarterclayton** [comment]
Call it priority two below getting the cluster part working.  Fork an issue to track them.

**Comment by pweil-** [comment]
@smarterclayton @pmorie @liggitt - admission is in a "good enough to review" state while I continue testing.  Integration with the allocation to annotations on namespaces is in.  Will continue to test and clean up but feel free to start taking a look.

/cc @abhgupta

**Comment by pweil-** [comment]
Now updated with a test case for admission that walks through:

1.  create an scc for the service accounts group
1.  test pods that should and shouldn't be admitted
1.  adding an scc with escalated privileges
1.  test that the pods that were previously denied are now accepted by the new scc

**Comment by smarterclayton** [comment]
For the end to end test case I want this enabled in uid range isolation mode - our system default can be run as non-root.

**Comment by pweil-** [comment]
@pmorie - I have one open issue we need to discuss.  When trying to validate for host volume plugins I found that the volume itself was not yet created so I was unable to retrieve it by the name on the pod and check for a host path.  I've commented that out of the provider for now.

@smarterclayton - ok, will create an e2e for this and make some good docs.  After that I will take a look at the kubelet changes required for a real non-root strategy (ie. docker inspect, set the uid if it is unspecified but the image def has it) as discussed earlier.  Will ping you tomorrow about how system defaults are best configured.

**Comment by pmorie** [comment]
I don't understand, @pweil-

**Comment by pweil-** [comment]
@pmorie when iterating through the volumeMounts on the container I have the name of a PV.  When I queried for that PV it was not found.  At first glance it looked to be the mount for the default token.  Specifically it is this code that is not working https://github.com/pweil-/origin/blob/523f4b0f9f86750e19e3903f737ea2fa27e23f17/Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/securitycontextconstraints/provider.go#L165-165

**Comment by pmorie** [comment]
Just look at the pod volumes, not persistent volumes for now. I need to do
a little background digging on PVs to understand how to look for them.

**Comment by liggitt** [comment]
@pweil- You can't look at the `Pod.Spec.Container[n].VolumeMounts`, you have to look at `Pod.Spec.Volumes` to determine the type of volumes the pod is using. Pods using HostDir will have a non-nil `VolumeSource.HostPathVolumeSource` field. Pods using persistent volumes will have a non-nil `VolumeSource.PersistentVolumeClaimVolumeSource` field. I'm not sure about the other types (NFS, EmptyDir, GCEPersistentDisk, ISCSI, etc).

**Comment by liggitt** [comment]
the name of a container's VolumeMount corresponds to the name of a Volume defined in the pod spec, nothing more

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/securitycontextconstraints/provider.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+				}
+			}
+			if !found {
+				allErrs = append(allErrs, fielderrors.NewFieldInvalid("capabilities.add", cap, "Capability is not allowed to be added"))
+			}
+		}
+	}
+
+	if !s.scc.AllowHostDirVolumePlugin {
+		// TODO: the volume is not yet available if we validate in admission, how can we validate this?
```
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/securitycontextconstraints/provider.go]: You want to look for:

1.  Whether `pod.Volumes` contains a `hostPath` volume (ie, volume.HostPath)
2.  Whether `pod.Volumes` contains a volume with a persistent volume claim source that refers to a `hostPath` persistent volume.

Btw, `scc.AllowHostDirVolumePlugin` should be `scc.AllowHostPathVolumePlugin`

File: pkg/securitycontext/admission/admission.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+	reflector.Run()
+
+	return &constraint{
+		Handler: kadmission.NewHandler(kadmission.Create, kadmission.Update),
+		client:  kclient,
+		store:   store,
+	}
+}
+
+func (c *constraint) Admit(a kadmission.Attributes) error {
```
Comment (liggitt) [inline pkg/securitycontext/admission/admission.go]: should this return if the action is not a "CREATE" action?

File: pkg/securitycontext/admission/admission.go [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+	//     2.  set the uid on the SCC
+	// 3.  Match the pod's SC to an SCC; for each SCC:
+	//     1.  generate the SC
+	//     2.  validate against the pod's SC
+
+	if a.GetResource() != string(kapi.ResourcePods) {
+		return nil
+	}
+
+	pod, ok := a.GetObject().(*kapi.Pod)
```
Comment (liggitt) [inline pkg/securitycontext/admission/admission.go]: not sure how this behaves with subresources... `GetResource()` may report pods for subresources which aren't of type *kapi.Pod... other admission controller return if they don't get the expected type rather than fail

File: pkg/securitycontext/admission/admission.go [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+	if !ok {
+		return errors.NewBadRequest("a pod was received, but could not convert the request object.")
+	}
+
+	// get the namespace
+	namespace, err := c.client.Namespaces().Get(a.GetNamespace())
```
Comment (liggitt) [inline pkg/securitycontext/admission/admission.go]: move this check later... the service account and scc fetches can be done with just the name. if you're not working from caches, we want to be as efficient as possible, since this adds several requests to every pod creation

File: pkg/securitycontext/admission/admission.go [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+	if err != nil {
+		return errors.NewBadRequest(err.Error())
```
Comment (liggitt) [inline pkg/securitycontext/admission/admission.go]: return admission errors, not generic errors (see other admission plugins for examples)

File: pkg/securitycontext/admission/admission.go [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+			if len(errs) != 0 {
+				continue
+			}
+
+			// Set the security context on actual container
+			pod.Spec.Containers[i].SecurityContext = context
+			if pod.Annotations == nil {
+				pod.Annotations = make(map[string]string, 0)
+			}
+			pod.Annotations[validatedSCCAnnotation] = providerName
```
Comment (liggitt) [inline pkg/securitycontext/admission/admission.go]: won't this overwrite the annotation if you get here for two containers in the same pod?

Comment (liggitt) [inline pkg/securitycontext/admission/admission.go]: what is making use of this annotation? should you start the `Admit()` method by ensuring it is cleared, so that it can only be set if we get here?

File: pkg/securitycontext/admission/admission.go [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+	return nil
+}
+
+// getMatchingSecurityContextConstraints returns constraints from the store that match the group,
+// uid, or user of the service account.
+func (c *constraint) getMatchingSecurityContextConstraints(ns string, serviceAccount *kapi.ServiceAccount) ([]*kapi.SecurityContextConstraints, error) {
+	userInfo := serviceaccount.UserInfo(ns, serviceAccount.Name, string(serviceAccount.UID))
+	matchedConstraints := make([]*kapi.SecurityContextConstraints, 0)
+
+	for _, c := range c.store.List() {
```
Comment (liggitt) [inline pkg/securitycontext/admission/admission.go]: `c` is masking the receiver... this could actually be a receiverless function

File: pkg/securitycontext/admission/admission.go [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+		// TODO: better error message
+		return errors.NewBadRequest("Unable to find a valid security context constraint for container")
+	}
+
+	return nil
+}
+
+// getMatchingSecurityContextConstraints returns constraints from the store that match the group,
+// uid, or user of the service account.
+func (c *constraint) getMatchingSecurityContextConstraints(ns string, serviceAccount *kapi.ServiceAccount) ([]*kapi.SecurityContextConstraints, error) {
```
Comment (liggitt) [inline pkg/securitycontext/admission/admission.go]: make this take a user.Info arg, then it can be reused to get constraints for a service account or the calling user

File: pkg/securitycontext/admission/admission.go [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+		}
+		for _, userGroup := range userInfo.GetGroups() {
+			if constraintSupportsGroup(userGroup, constraint.Groups) {
+				matchedConstraints = append(matchedConstraints, constraint)
+				break
+			}
+		}
+
+		for _, user := range constraint.Users {
+			if userInfo.GetName() == user || userInfo.GetUID() == user {
```
Comment (liggitt) [inline pkg/securitycontext/admission/admission.go]: only compare GetName()

File: pkg/securitycontext/admission/admission.go [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+			}
+		}
+	}
+
+	return matchedConstraints, nil
+}
+
+// getPreallocatedUIDRange retrieves the annotated value from the service account, splits it to make
+// the min/max and formats the data into the necessary types for the strategy options.
+func getPreallocatedUIDRange(ns *kapi.Namespace) (min *int64, max *int64, err error) {
```
Comment (liggitt) [inline pkg/securitycontext/admission/admission.go]: don't use named return args... makes it hard to follow what is actually being returned

File: pkg/securitycontext/admission/admission.go [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+	if e != nil {
+		err = e
+		return
+	}
+	max = &iMax
+	return
+}
+
+// getPreallocatedLevel gets the annotated value from the service account and ensures that
+// s0 is added.
```
Comment (liggitt) [inline pkg/securitycontext/admission/admission.go]: a super-brief comment explaining why would be appreciated

File: pkg/securitycontext/admission/admission.go [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+	}
+
+	// all pods are required to have a service account
+	serviceAccount, err := getServiceAccount(c.client, namespace.Name, pod)
+	if err != nil {
+		return errors.NewBadRequest(err.Error())
+	}
+
+	// get all constraints that are usable by this service account
+	matchedConstraints, err := c.getMatchingSecurityContextConstraints(namespace.Name, serviceAccount)
```
Comment (liggitt) [inline pkg/securitycontext/admission/admission.go]: is getting the allowed SCC's for the calling user going to be a follow-up?

File: pkg/securitycontext/admission/admission.go (hunk 2 above)
Comment (pweil-) [inline pkg/securitycontext/admission/admission.go]: This uses the Handler above to ignore actions that it does not support.  In the chain handler there is a check with the .Handles method so I don't think this is necessary.

File: pkg/securitycontext/admission/admission.go (hunk 12 above)
Comment (pweil-) [inline pkg/securitycontext/admission/admission.go]: Yes, we are operating under the assumption here that every pod will run with a service account.

File: pkg/securitycontext/admission/admission.go (hunk 6 above)
Comment (pweil-) [inline pkg/securitycontext/admission/admission.go]: hrm, that's true.  Perhaps this should be formatted with the container name in it.  This is for two reasons, telling at runtime what allowed the pod to run and also to establish the pattern of how we can tell the kubelet that it was validated with a strategy that requires kubelet interaction (the non-root strategy)

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/securitycontextconstraints/provider.go (hunk 1 above)
Comment (pweil-) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/securitycontextconstraints/provider.go]: Yes, that's what I'm shooting for but the issue I'm having is that I only have the name of the volume on the pod spec.  Retrieving the volume to do the check is what is failing with not found.

**Comment by pweil-** [comment]
> the name of a container's VolumeMount corresponds to the name of a Volume defined in the pod spec, nothing more

ah ha...got it.  Thanks @liggitt @pmorie

File: pkg/securitycontext/admission/admission.go (hunk 11 above)
Comment (pweil-) [inline pkg/securitycontext/admission/admission.go]: reviewing the label allocator code again I think this may be altogether wrong.  Will update.

File: pkg/securitycontext/admission/admission.go (hunk 12 above)
Comment (liggitt) [inline pkg/securitycontext/admission/admission.go]: if we don't also allow SCC's for the calling user, the build controller won't be able to create privileged pods unless the service account is also given that permission, which we didn't want to do

**Comment by pweil-** [comment]
ok, updates based on most feedback are now in.  @smarterclayton FYI, I changed master.go to add the `s0` level.  Confirmed with @eparis that it required.

TODO: address @liggitt's comment about users SCCs.  @smarterclayton - @pmorie and I were operating under the assumption that all pods would run with a service account.  Things like deployments/builds would have an SA that has a privileged SCC.  True/false?

TODO: parse out the mcs prefix and set it into SELinuxOptions if it has user/type/role

PTAL

**Comment by liggitt** [comment]
>  Things like deployments/builds would have an SA that has a privileged SCC. True/false?

False. We shouldn't give service accounts in every namespace permission to run privileged pods.

**Comment by liggitt** [comment]
Deployment pods (will) run using a "deployer" service account in the local namespace. Build pods do run using a "builder" service account in the local namespace. Neither of those service accounts should be able to launch privileged pods on their own (or every project admin/editor will also be able to do that)

**Comment by pweil-** [comment]
@smarterclayton @liggitt addressed feedback

File: pkg/security/admission/admission.go [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+	}
+
+	pod, ok := a.GetObject().(*kapi.Pod)
+	// if we can't convert then we don't handle this object so just return
+	if !ok {
+		return nil
+	}
+
+	// all pods are required to have a service account
+	serviceAccount, err := getServiceAccount(c.client, a.GetNamespace(), pod)
```
Comment (liggitt) [inline pkg/security/admission/admission.go]: if there's no service account, shouldn't it just use the constraints for the calling user? could even check if len(pod.spec.serviceaccount) > 0 first. keep mirror pods in mind as well... those won't have service accounts and mutating their specs will make the kubelet angry

Comment (pweil-) [inline pkg/security/admission/admission.go]: Yes, you're right this case is not a deal breaker and shouldn't fail hard.

**Comment by pweil-** [comment]
@pmorie @liggitt @smarterclayton - updated with SCCs to use during e2e and test-cmd.  However, while debugging I noticed that the users/groups being used by a lot of the items were not specific enough to prevent creation of things like the router to be confined properly.  I ended up needing an SCC that looks like this:

```json
{
  "kind": "SecurityContextConstraints",
  "apiVersion":"v1beta3",

  "metadata": {
    "name": "scc-admin",
    "labels": {
      "name": "scc-admin"
    }
  },

  "allowPrivilegedContainer": true,
  "allowHostDirVolumePlugin": true,
  "runAsUser": {
    "type": "RunAsAny"
  },
  "seLinuxContext": {
    "type": "RunAsAny"
  },
  "users": ["system:openshift-client", "system:serviceaccount:default:default"],
  "groups": ["system:nodes", "system:deployers", "system:cluster-admins"]
}
```
which is NOT what we want.  Perhaps the policy experts can help me out here. 

Sorting through who needs what seems like it will be the next big task.  I ran into:

1.  the registry mounts a hostdir and my admin account hadn't specified hostdir=true.  It will need an SCC with hostdir=true
1.  the router needs to access directories in the image that are owned by root so it needs a RunAsUser allowing SCC
1.  issues with things being created as `system:openshift-client` and the default SAs.  Some debug logging: https://gist.github.com/pweil-/b6b3bf6b400aa241e026

**Comment by smarterclayton** [comment]
> On Jun 2, 2015, at 8:29 PM, Paul Weil <notifications@github.com> wrote:
> 
> @pmorie @liggitt @smarterclayton - updated with SCCs to use during e2e and test-cmd. However, while debugging I noticed that the users/groups being used by a lot of the items were not specific enough to prevent creation of things like the router to be confined properly. I ended up needing an SCC that looks like this:
> 
> {
>   "kind": "SecurityContextConstraints",
>   "apiVersion":"v1beta3",
> 
>   "metadata": {
>     "name": "scc-admin",
>     "labels": {
>       "name": "scc-admin"
>     }
>   },
> 
>   "allowPrivilegedContainer": true,
>   "allowHostDirVolumePlugin": true,
>   "runAsUser": {
>     "type": "RunAsAny"
>   },
>   "seLinuxContext": {
>     "type": "RunAsAny"
>   },
>   "users": ["system:openshift-client", "system:serviceaccount:default:default"],
>   "groups": ["system:nodes", "system:deployers", "system:cluster-admins"]
> }
Why would system:nodes need an SCC?
> which is NOT what we want. Perhaps the policy experts can help me out here.
> 
> Sorting through who needs what seems like it will be the next big task. I ran into:
> 
> the registry mounts a hostdir and my admin account hadn't specified hostdir=true. It will need an SCC with hostdir=true
It needs to stop doing this.
> the router needs to access directories in the image that are owned by root so it needs a RunAsUser allowing SCC
The router runs as root today, but it should be running as non-root but with the right caps for the port.  The directory it writes to should not be writable only by root.

**Comment by pmorie** [comment]
> > the registry mounts a hostdir and my admin account hadn't specified hostdir=true. It will need an SCC with hostdir=true

> It needs to stop doing this.

+1 to this -- we shouldn't be using the cli option to do the hostpath mount for the e2e tests.

**Comment by pweil-** [comment]
> Why would system:nodes need an SCC?

I believe it was caused by pod status updates going through the admission controller.  I ran into a scenario where the builder was stuck in pending even when it was actually complete.  This is probably an indicator for some improvements in admission so I wanted to point it out.  An example of one of the items that came in under `system:nodes`:

```
scc: pod database-1-deploy is using user &{system:node-127.0.0.1  [system:nodes system:authenticated]}
```

> The router runs as root today, but it should be running as non-root but with the right caps for the port.  The directory it writes to should not be writable only by root.

yes

**Comment by liggitt** [comment]
> I believe it was caused by pod status updates going through the admission controller

The admission controller should only be processing CREATE calls if the pod spec is immutable post-create

**Comment by derekwaynecarr** [comment]
Right, pod status changes is a pod update.  looks like you do not need update in admission handler

**Comment by smarterclayton** [comment]
> On Jun 2, 2015, at 9:50 PM, Paul Weil <notifications@github.com> wrote:
> 
> Why would system:nodes need an SCC?
> 
> I believe it was caused by pod status updates going through the admission controller. I ran into a scenario where the builder was stuck in pending even when it was actually complete. This is probably an indicator for some improvements in admission so I wanted to point it out. An example of one of the items that came in under system:nodes:
> 
You shouldn't be running your admission controller for pods/status.  It can't change anything related to security.

**Comment by liggitt** [comment]
> You shouldn't be running your admission controller for pods/status. It can't change anything related to security.

I think this was a bug in admission and subresources... the kind is still "pods" and the object is a *api.Pod... there's no way to tell it's pods/status

**Comment by derekwaynecarr** [comment]
Not all of the pod spec is immutable.  You can change some parts of the container.  

https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/validation/validation.go#L988

**Comment by smarterclayton** [comment]
I though Derek and I talked about this...

**Comment by liggitt** [comment]
yeah, image, but nothing security-context-related, right?

**Comment by derekwaynecarr** [comment]
An update of a pod is an update to pod status or pod spec.  Admission control is fine.  Pod spec allows updating container piece excluding the image name.  Admission on update could get old version and compare with incoming change.

**Comment by liggitt** [comment]
oh, ok... I was expecting to be able to see the subresourced kind

**Comment by pweil-** [comment]
will remove update from the handler definition

**Comment by liggitt** [comment]
what was `system:deployers` needing?

**Comment by derekwaynecarr** [comment]
Clayton we chatted about changing the resource to be pods/status

https://github.com/GoogleCloudPlatform/kubernetes/issues/8868

But that is not done yet

**Comment by derekwaynecarr** [comment]
Sub resource will appear but I thought it was most likely post 3.0 given code freeze.  I can put a PR up tomorrow to fix upstream issue and see the response

**Comment by liggitt** [comment]
not a big deal if we don't need update here

**Comment by smarterclayton** [comment]
It doesn't have to be 3.0 if we can't update SCC anyway

**Comment by pweil-** [comment]
> what was system:deployers needing?

I was debugging for too long and don't remember specifically what was failing.  I'm going to run through e2e again with update removed to reduce the list.  Will report back.

**Comment by liggitt** [comment]
Also, the `system:deployers` group will disappear in https://github.com/openshift/origin/pull/2530... each namespace will have it's own deployer service account

**Comment by pweil-** [comment]
@liggitt @smarterclayton @pmorie  findings from my work so far today

1. registry with no host mount uses an empty dir which is 750 owned by root (worked around in testing by pausing e2e, nsenter + chmod 777 /registry) - when that is fixed then we should be able to run without an admin SCC
1.  router will work with a non-admin SCC by changing the permission on /var (for now, empty dir with fixed permissions in the future) in the docker file and adding `setcap 'cap_net_bind_service=+ep' /usr/sbin/haproxy` @eparis - thoughts on this approach?
1.  the sti builder in e2e is requesting a privileged container and mounting a host volume (info below)
1.  the exec command on the non-root registry pod will fail because whoami can't return a username, I changed it to id with grep 10000 which would is a substring of the range of the contiguous allocator

```
scc: using user info &{system:openshift-client  [system:authenticated]} for pod ruby-sample-build-1-build
I0603 20:03:38.686828    9790 admission.go:99] scc: using sa user info &{system:serviceaccount:test:default d4457111-0a2a-11e5-b958-080027c5bfa9 [system:serviceaccounts system:serviceaccounts:test]} for pod ruby-sample-build-1-build
I0603 20:03:38.686857    9790 admission.go:158] scc: found providers map[string]securitycontextconstraints.SecurityContextConstraintsProvider{"scc-serviceaccounts":(*securitycontextconstraints.simpleProvider)(0xc20eacde60)} for pod ruby-sample-build-1-build
I0603 20:03:38.687413    9790 admission.go:202] provider scc-serviceaccounts produced errors [privileged: invalid value 'true': Privileged containers are not allowed VolumeMounts: invalid value 'docker-socket': Host Volumes are not allowed to be used] for pod sti-build container %!s(MISSING)
```

I ended up with a successful test with the steps above and SCCs of 

```
{
  "kind": "SecurityContextConstraints",
  "apiVersion":"v1beta3",

  "metadata": {
    "name": "scc-admin",
    "labels": {
      "name": "scc-admin"
    }
  },

  "allowPrivilegedContainer": true,
  "allowHostDirVolumePlugin": true,
  "runAsUser": {
    "type": "RunAsAny"
  },
  "seLinuxContext": {
    "type": "RunAsAny"
  },
  "users": [],
  "groups": ["system:cluster-admins", "system:serviceaccounts:test"]
}

{
  "kind": "SecurityContextConstraints",
  "apiVersion":"v1beta3",

  "metadata": {
    "name": "scc-serviceaccounts",
    "labels": {
      "name": "scc-serviceaccounts"
    }
  },
  "runAsUser": {
    "type": "MustRunAsRange"
  },
  "seLinuxContext": {
    "type": "MustRunAs"
  },
  "groups": ["system:serviceaccounts"]
}
```

**Comment by liggitt** [comment]
do you also want all calling users to get the restrictive SCC by default? if so, add 
```
"groups": ["system:serviceaccounts", "system:authenticated"]
```

Builds need to work in any namespace, not just in the test namespace. that means we need to give the buildcontroller user (the one creating the build pods) privileged access. Right now, that's `system:openshift-client`, but we can't give them privileged access, since they also run the replication controller, which would allow any pod anywhere to be privileged. Sounds like we need a unique user for the build controller.

**Comment by pweil-** [comment]
> do you also want all calling users to get the restrictive SCC by default? if so, add

Adding

**Comment by liggitt** [comment]
also, why the labels on the SCC objects? just curious

**Comment by pweil-** [comment]
> Sounds like we need a unique user for the build controller.

I agree with this.  I think the open issues are 

1.  How to identify the builds
1.  Should we provide default SCCs at startup or consider this an admin action to get the cluster running 
1.  If it's an admin action should we add a command to easily produce the known set SCCs for a setup (like the two used in e2e)
1.  Run as non-root (identified as lower priority and tracked on a follow up issue)

> also, why the labels on the SCC objects? just curious

No particular reason, copy pasta from another template

**Comment by smarterclayton** [comment]
----- Original Message -----
> @liggitt @smarterclayton @pmorie  findings from my work so far today
> 
> 1. registry with no host mount uses an empty dir which is 750 owned by root
> (worked around in testing by pausing e2e, nsenter + chmod 777 /registry) -
> when that is fixed then we should be able to run without an admin SCC
> 1.  router will work with a non-admin SCC by changing the permission on /var
> (for now, empty dir with fixed permissions in the future) in the docker file
> and adding `setcap 'cap_net_bind_service=+ep' /usr/sbin/haproxy` @eparis -
> thoughts on this approach?
> 1.  the sti builder in e2e is requesting a privileged container and mounting
> a host volume (info below)
> 1.  the exec command on the non-root registry pod will fail because whoami
> can't return a username, I changed it to id with grep 10000 which would is a
> substring of the range of the contiguous allocator

This is what the nsswitch.conf plugin that would like about the id was supposed to do - @ncdc @csrwng

**Comment by pweil-** [comment]
Spawned some related issues for tracking since they are not directly related to the SCC but need addressed for full integration

1.  Builders: https://github.com/openshift/origin/issues/2774
1.  Cluster defaults: https://github.com/openshift/origin/issues/2775
1.  Non-root follow up: https://github.com/openshift/origin/issues/2626
1.  Empty dir: https://github.com/GoogleCloudPlatform/kubernetes/issues/2630

**Comment by smarterclayton** [comment]
> On Jun 3, 2015, at 5:11 PM, Paul Weil <notifications@github.com> wrote:
> 
> Sounds like we need a unique user for the build controller.
> 
> I agree with this. I think the open issues are 
> 
> How to identify the builds
Why do they need to be identified?
> Should we provide default SCCs at startup or consider this an admin action to get the cluster running
If the system isn't secure without them, they need to be part of policy bootstrap

**Comment by pweil-** [comment]
> Why do they need to be identified?

because they require escalated privileges (ie sti builder needs host mounts and privileged) but are currently identified with users/groups that would affect items we don't want to be run with escalated privileges.  The currently run with `system:openshift-client` as does replication controllers

**Comment by liggitt** [comment]
>> How to identify the builds
>
> Why do they need to be identified?

How to identify the build pod creator

**Comment by smarterclayton** [comment]
That's what I was looking for.

**Comment by pweil-** [comment]
breaking this up into smaller PRs so we can start landing it.

**Comment by pweil-** [comment]
Closing this in favor of 

Upstream: https://github.com/openshift/origin/pull/2856
Default cluster SCC: https://github.com/openshift/origin/pull/2863
Admission: https://github.com/openshift/origin/pull/2868
Build Pod Creators: https://github.com/openshift/origin/issues/2774
Empty Dir Permissions: https://github.com/GoogleCloudPlatform/kubernetes/issues/2630
Follow up for RunAsNonRoot strategy: https://github.com/openshift/origin/issues/2626
